### PR TITLE
reimplement error handling: use anyhow

### LINKED ISSUE #641: agent: use anyhow for error handling (remove error_chain and custom errors)
currently custom errors/error_chain are used for error handling, however those implementations are no thread safe.
anyhow is easier to use and thread safe.

### PR DESCRIPTION
**Pull request by devimc**
Don't use `rustjail::errors` for error handling, since it's not
thread safe and there are better alternatives like `anyhow`.

`anyhow` attaches context to help the person troubleshooting
the error understand where things went wrong, for example:

Current error messages:

```
No such file or directory (os error 2)
```

With `anyhow`:

```
Error: Failed to read config.json
Caused by:
    No such file or directory (os error 2)
```
    
fixes #641

### CODE REVIEW

**Comment by devimc** [comment]
@lifupan @bergwolf @liubin  please take a look

File: src/agent/rustjail/src/cgroups/fs/mod.rs [hunk 1]
```diff
@@ -313,47 +321,45 @@ where
 {
     let p = format!("{}/{}", dir, file);
     fs::write(p.as_str(), v.to_string().as_bytes())
-        .chain_err(|| format!("couldn't write to file: {:?}", p))?;
+        .map_err(|e| anyhow!(e).context(format!("couldn't write to file: {:?}", p)))?;
```
Comment (lifupan) [inline src/agent/rustjail/src/cgroups/fs/mod.rs:324]: why map_err here?  why  not using  context directly such as: 

fs::write(p.as_str(), v.to_string().as_bytes()).context(format!("couldn't write to file: {:?}", p))?;

File: src/agent/rustjail/src/cgroups/fs/mod.rs [hunk 2]
```diff
@@ -313,47 +321,45 @@ where
...
+
     Ok(())
 }
 
 fn read_file(dir: &str, file: &str) -> Result<String> {
     let p = format!("{}/{}", dir, file);
-    let ret = fs::read_to_string(p.as_str())?;
+    let ret = fs::read_to_string(p.as_str()).map_err(|e| anyhow!(e))?;
```
Comment (lifupan) [inline src/agent/rustjail/src/cgroups/fs/mod.rs:331]: Same here.

File: src/agent/rustjail/src/cgroups/fs/mod.rs [hunk 3]
```diff
@@ -313,47 +321,45 @@ where
...
-        return Err(ErrorKind::ErrorCode("cannot copy file from parent".to_string()).into());
+        return Err(anyhow!("cannot copy file from parent"));
     };
 
     match read_file(parent, file) {
         Ok(v) => {
             if !v.trim().is_empty() {
                 info!(sl!(), "value: \"{}\"", v.as_str().trim());
-                return write_file(dir, file, v);
+                return write_file(dir, file, v).map_err(|e| anyhow!(e));
```
Comment (lifupan) [inline src/agent/rustjail/src/cgroups/fs/mod.rs:346]: same here.

File: src/agent/rustjail/src/cgroups/fs/mod.rs [hunk 4]
```diff
@@ -313,47 +321,45 @@ where
...
+            return copy_parent(dir, file);
         }
         Err(e) => return Err(e.into()),
     }
 }
 
 fn write_nonzero(dir: &str, file: &str, v: i128) -> Result<()> {
     if v != 0 {
-        write_file(dir, file, v.to_string())?;
+        write_file(dir, file, v.to_string()).map_err(|e| anyhow!(e))?;
```
Comment (lifupan) [inline src/agent/rustjail/src/cgroups/fs/mod.rs:362]: The same here.

**Comment by lifupan** [comment]
> @lifupan @bergwolf @liubin please take a look

Hi @devimc 

Since anyhow would convert those std  Error into anyhow Error type  automatically, thus I think  you can  remove those map_err converting  in the patches.

File: src/agent/rustjail/src/cgroups/fs/mod.rs (hunk 1 above)
Comment (devimc) [reply src/agent/rustjail/src/cgroups/fs/mod.rs:324]: didn't work

```
error[E0599]: no method named `context` found for enum `std::result::Result<(), std::io::Error>` in the current scope
   --> rustjail/src/cgroups/fs/mod.rs:324:10
    |
324 |         .context(format!("couldn't write to file: {:?}", p))?;
    |          ^^^^^^^ method not found in `std::result::Result<(), std::io::Error>`
```

File: src/agent/rustjail/src/cgroups/fs/mod.rs (hunk 2 above)
Comment (devimc) [reply src/agent/rustjail/src/cgroups/fs/mod.rs:331]: same here

File: src/agent/rustjail/src/cgroups/fs/mod.rs (hunk 3 above)
Comment (devimc) [reply src/agent/rustjail/src/cgroups/fs/mod.rs:346]: fixed

File: src/agent/rustjail/src/cgroups/fs/mod.rs (hunk 4 above)
Comment (devimc) [reply src/agent/rustjail/src/cgroups/fs/mod.rs:362]: fixed

**Comment by devimc** [comment]
thanks @lifupan , PR  updated

File: src/agent/rustjail/src/cgroups/fs/mod.rs (hunk 1 above)
Comment (lifupan) [reply src/agent/rustjail/src/cgroups/fs/mod.rs:324]: Hi @devimc 

It seems you should import  anyhow's Context  trait  to use context method such as:
use anyhow::{anyhow, Context, Result};

Comment (devimc) [reply src/agent/rustjail/src/cgroups/fs/mod.rs]: you're right, sorry this is my second patch in rust :disappointed:

**Comment by devimc** [comment]
@lifupan fixed, thanks

**Review by lifupan — APPROVED** [review]

**Comment by devimc** [comment]
@fidencio @bergwolf @amshinde please take a look

**Review by c3d — COMMENTED** [review]
I would slightly rephrase the comment to indicate what you expect from this.

My guess is that the main benefit is to be able to give a context for an error. Maybe an illustration showing the before/after errors?

**Comment by devimc** [comment]
PR and comment updated

/test

**Review by jodh-intel — APPROVED** [review]
Nice! Thanks @devimc!

lgtm

File: src/agent/rustjail/src/mount.rs [hunk 5]
```diff
@@ -139,7 +139,7 @@ pub fn init_rootfs(
     for m in &spec.mounts {
         let (mut flags, data) = parse_mount(&m);
         if !m.destination.starts_with("/") || m.destination.contains("..") {
-            return Err(ErrorKind::Nix(nix::Error::Sys(Errno::EINVAL)).into());
+            return Err(anyhow!(nix::Error::Sys(Errno::EINVAL)));
```
Comment (c3d) [inline src/agent/rustjail/src/mount.rs:142]: Curious what is the difference between `nix::Error::Sys(Errno::EINVAL))` and `anyhow!(nix::Error::from_errno(Errno::EINVAL)` (used above). From reading the doc, I see no real difference.

If there is no real difference, consider making the code consistent between the two locations.

**Review by c3d — APPROVED** [review]
Overall looks good. A few messages should be rephrased, but I see that this has been merged already… So that will come later.

**Pull request merged**